File: examples/awfy/node/queens.mjs
export const warmup = 10, iters = 40, size = 1000, expected = 1000;

function getRowColumn(freeRows, freeMaxs, freeMins, r, c) {
  return freeRows[r] && freeMaxs[c + r] && freeMins[c - r + 7];
}

function setRowColumn(freeRows, freeMaxs, freeMins, r, c, v) {
  freeRows[r] = v;
  freeMaxs[c + r] = v;
  freeMins[c - r + 7] = v;
}

function place(freeRows, freeMaxs, freeMins, c) {
  if (c === 8) return true;
  for (let r = 0; r < 8; r++) {
    if (getRowColumn(freeRows, freeMaxs, freeMins, r, c)) {
      setRowColumn(freeRows, freeMaxs, freeMins, r, c, false);
      if (place(freeRows, freeMaxs, freeMins, c + 1)) return true;
      setRowColumn(freeRows, freeMaxs, freeMins, r, c, true);
    }
  }
  return false;
}

export function run(size) {
  let count = 0;
  for (let n = 0; n < size; n++) {
    const freeRows = new Array(8).fill(true);
    const freeMaxs = new Array(15).fill(true);
    const freeMins = new Array(15).fill(true);
    if (place(freeRows, freeMaxs, freeMins, 0)) count++;
  }
  return count;
}

File: examples/performance/awfy/go/queens.go
package main

func queensGetRowColumn(freeRows, freeMaxs, freeMins []bool, r, c int) bool {
	return freeRows[r] && freeMaxs[c+r] && freeMins[c-r+7]
}

func queensSetRowColumn(freeRows, freeMaxs, freeMins []bool, r, c int, v bool) {
	freeRows[r] = v
	freeMaxs[c+r] = v
	freeMins[c-r+7] = v
}

func queensPlace(freeRows, freeMaxs, freeMins []bool, c int) bool {
	if c == 8 {
		return true
	}
	for r := 0; r < 8; r++ {
		if queensGetRowColumn(freeRows, freeMaxs, freeMins, r, c) {
			queensSetRowColumn(freeRows, freeMaxs, freeMins, r, c, false)
			if queensPlace(freeRows, freeMaxs, freeMins, c+1) {
				return true
			}
			queensSetRowColumn(freeRows, freeMaxs, freeMins, r, c, true)
		}
	}
	return false
}

func fillTrue(n int) []bool {
	s := make([]bool, n)
	for i := range s {
		s[i] = true
	}
	return s
}

func queensRun(size int) int {
	count := 0
	for n := 0; n < size; n++ {
		freeRows := fillTrue(8)
		freeMaxs := fillTrue(15)
		freeMins := fillTrue(15)
		if queensPlace(freeRows, freeMaxs, freeMins, 0) {
			count++
		}
	}
	return count
}

var queensBench = Bench{Name: "queens", Warmup: 10, Iters: 40, Size: 1000, Expected: 1000, Run: queensRun}
